confmap/provider/googlesecretsprovider/provider.go

```diff
@@ -0,0 +1,78 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//go:generate mdatagen metadata.yaml
+
+package googlesecretsprovider // import "github.com/open-telemetry/opentelemetry-collector-contrib/confmap/provider/googlesecretsprovider"
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	secretmanager "cloud.google.com/go/secretmanager/apiv1"
+	gax "github.com/googleapis/gax-go/v2"
+	"github.com/googleapis/gax-go/v2/apierror"
+	"go.opentelemetry.io/collector/confmap"
+
+	"cloud.google.com/go/secretmanager/apiv1/secretmanagerpb"
+)
+
+type secretsManagerClient interface {
+	AccessSecretVersion(ctx context.Context, req *secretmanagerpb.AccessSecretVersionRequest, opts ...gax.CallOption) (*secretmanagerpb.AccessSecretVersionResponse, error)
+	Close() error
+}
+
+const (
+	schemeName = "googlesecretsprovider"
+)
+
+type provider struct {
+	client secretsManagerClient
+}
+
+func NewFactory() confmap.ProviderFactory {
+	return confmap.NewProviderFactory(newProvider)
+}
+
+func newProvider(confmap.ProviderSettings) confmap.Provider {
+	return &provider{client: nil}
+}
+
+func (p *provider) Retrieve(ctx context.Context, uri string, _ confmap.WatcherFunc) (*confmap.Retrieved, error) {
+	if !strings.HasPrefix(uri, schemeName+":") {
+		return nil, fmt.Errorf("%q uri is not supported by %q provider", uri, schemeName)
+	}
+	secretName := strings.TrimPrefix(uri, schemeName+":")
+
+	if p.client == nil {
+		client, err := secretmanager.NewClient(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create a Google secret manager client: %w", err)
+		}
+		defer client.Close()
```

Comment:
Wouldn't this cause a failure if you tried to request more than one secret? The client would be not nil but closed after the first request. Maybe the client should be closed on `Shutdown` if it's not nil.

Reply:
I think Retrieve() is called sequentially: https://github.com/open-telemetry/opentelemetry-collector/blob/c85ebbec779fa00c3fbe4a555eb0d420bdad7581/confmap/resolver.go#L177. 

Also note: https://github.com/open-telemetry/opentelemetry-collector/blob/c85ebbec779fa00c3fbe4a555eb0d420bdad7581/confmap/resolver.go#L167

However, the secret manager client connection can be reused across multiple calls to Retrieve, so I've moved the `Close` call to Shutdown. I also added a lock, in case of any race conditions.